### ISSUE TITLE
Add `ninja -t outputs` to print modified targets

### ISSUE THREAD

**Issue by Danoloan10** [ninja-build/ninja#2133]
A new tool that could be useful is a `-t outputs` tool that:
- Given a list of source files (stdin? file specified in option?) prints what targets would be built if they were modified.
- If no source file list is given, it could print what targets would be built if given the target argument
- An option `--leaf` to print only leaf targets

Full command line:
```
ninja -t outputs [-f source_list] [-l|--leaf] [target]
```

# Rationale
CMake integrates the test execution framework CTest, but it lacks support to selectively compile and run tests when some files are modified. This tool would not only make it easy to bypass this issue, but would expose the power of dependency resolution Ninja has to third-party applications that may find it useful.

# Alternative
A good alternative to adding a new tool would be to expose this functionality through a Python API. Well, it would be a good idea to expose all tools and most ninja functionalities through a Python API. I don't know if this is already done because I couldn't find the proper documentation about it.

Please feel free to comment whether this functionality can be achieved already by using some tool/cli/configuration I may be unaware of.

**Comment by jonesmz** [comment]
If Ninja were a C++ library (instead of a program), with a foreign-function-interface wrapper for python to call into, would that satisfy what you're looking for here?

**Comment by Danoloan10** [comment]
If that library would expose the functionality that I've described (being able to get the list of targets that transitively depend on a list of sources, and defaulting to only modified files if the list is empty), that would be enough to be able to make third party applications that interface with and extend ninja, like the example I mentioned with CTest.

The bypass I have thought of now involves executing `ninja -t query` with a breadth-first search algorithm. This could give the targets given the sources as inputs. However, it would exec ninja a lot and would likely be very inefficient.

**Comment by jhasse** [comment]
What you want is to have your tests be part of the build graph. My suggestion would be to write that logic yourself and drop CTest. CTest sucks.

**Comment by Danoloan10** [comment]
CTest sucks a bit yeah... but through conventions you can map test names with actual build targets. The problem here is to get what build targets depend on given a list of files, without the files being updated nor actually buiding the targets. This must be done to know what tests must be **run** when X files have been modified whether CTest is being used or not. Given that ninja is the software that already implements this logic (dependency graph definition & resolution), implementing it again would be redundant.

**Comment by jhasse** [comment]
> implementing it again would be redundant.

Exactly. That's why it's best to drop CTest and just use a custom build step for tests.

**Comment by marcelolynch** [comment]
> If Ninja were a C++ library (instead of a program), with a foreign-function-interface wrapper for python to call into, would that satisfy what you're looking for here?

@jonesmz - is something like this possible / something you would consider if a PR came in? I would be very interested in a way to navigate the graph after Ninja constructs it in memory, but without running the build - I have implemented a `RUN_AFTER_LOAD` tool in a Ninja fork for this purpose but an "official" way to do this instead would be great.

(Sorry for hijacking this issue, I can open another one if this merits discussion)

**Comment by jhasse** [comment]
This has been discussed before, see #1499.

My stance can be summed up with "there are other issues I'm more interested in and I don't have enough time left to review the changes".
